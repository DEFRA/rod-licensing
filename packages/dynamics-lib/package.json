{
  "name": "@defra-fish/dynamics-lib",
  "version": "1.22.0-rc.3",
  "description": "Framework to support integration with dynamics",
  "type": "module",
  "engines": {
    "node": ">=14"
  },
  "keywords": [
    "rod",
    "licensing",
    "dynamics"
  ],
  "author": "DEFRA",
  "license": "SEE LICENSE IN LICENSE",
  "main": "src/index.js",
  "directories": {
    "lib": "src",
    "test": "__tests__"
  },
  "files": [
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "git://github.com/DEFRA/rod-licensing.git",
    "directory": "packages/dynamics-lib"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "dependencies": {
    "cache-manager": "^3.4.3",
    "cache-manager-ioredis": "^2.1.0",
    "debug": "^4.3.1",
    "dynamics-web-api": "1.7.3",
    "joi": "^17.4.0",
    "moment": "^2.29.1",
    "pluralize": "^8.0.0",
    "simple-oauth2": "^4.2.0",
    "uuid": "^8.3.2"
  }
}
