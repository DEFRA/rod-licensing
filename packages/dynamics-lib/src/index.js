// Entities
export * from './entities/contact.entity.js'
export * from './entities/permission.entity.js'
export * from './entities/permit.entity.js'
export * from './entities/permit-concession.entity.js'
export * from './entities/concession.entity.js'
export * from './entities/concession-proof.entity.js'
export * from './entities/fulfilment-request.entity.js'
export * from './entities/fulfilment-request-file.entity.js'
export * from './entities/transaction.entity.js'
export * from './entities/transaction-journal.entity.js'
export * from './entities/transaction-currency.entity.js'
export * from './entities/pocl-file.entity.js'
export * from './entities/pocl-staging-exception.entity.js'
export * from './entities/pocl-data-validation-error.entity.js'
export * from './entities/recurring-payment.entity.js'
export * from './entities/recurring-payment-instruction.entity.js'
export * from './entities/staging-exception.entity.js'

export * from './entities/role.entity.js'
export * from './entities/system-user-role.entity.js'
export * from './entities/system-user.entity.js'

// Queries
export * from './queries/permission.queries.js'
export * from './queries/fulfilment.queries.js'

// Framework functionality
export * from './client/util.js'
export { dynamicsClient } from './client/dynamics-client.js'
export * from './client/cache.js'
export * from './client/entity-manager.js'
