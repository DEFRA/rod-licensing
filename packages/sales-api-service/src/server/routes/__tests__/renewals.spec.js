import Boom from '@hapi/boom'
import permissionRenewalData from '../renewals.js'
import { preparePermissionDataForRenewal } from '../../../services/renewals/renewals.service.js'
import { executeQuery, permissionForFullReferenceNumber } from '@defra-fish/dynamics-lib'
import {
  MOCK_EXISTING_PERMISSION_ENTITY,
  MOCK_EXISTING_CONTACT_ENTITY,
  MOCK_1DAY_SENIOR_PERMIT_ENTITY,
  MOCK_CONCESSION_PROOF_ENTITY,
  MOCK_CONCESSION
} from '../../../__mocks__/test-data.js'

jest.mock('../../../services/renewals/renewals.service.js', () => ({
  preparePermissionDataForRenewal: jest.fn()
}))

jest.mock('@defra-fish/dynamics-lib', () => ({
  ...jest.requireActual('@defra-fish/dynamics-lib'),
  permissionForFullReferenceNumber: jest.fn(),
  executeQuery: jest.fn()
}))

const getMockRequest = (referenceNumber = 'ABC123') => ({
  params: { referenceNumber }
})

const getMockResponseToolkit = () => ({
  response: jest.fn()
})

const permissionForFullReferenceNumberMock = () => ({
  entity: MOCK_EXISTING_PERMISSION_ENTITY,
  expanded: {
    licensee: { entity: MOCK_EXISTING_CONTACT_ENTITY, expanded: {} },
    concessionProofs: [{ entity: MOCK_CONCESSION_PROOF_ENTITY, expanded: { concession: { entity: MOCK_CONCESSION } } }],
    permit: { entity: MOCK_1DAY_SENIOR_PERMIT_ENTITY, expanded: {} }
  }
})

describe('permissionRenewalData', () => {
  beforeEach(jest.clearAllMocks)

  it('should call permissionForFullReferenceNumber with referenceNumber', async () => {
    executeQuery.mockResolvedValueOnce([permissionForFullReferenceNumberMock()])

    const referenceNumber = 'REFERENCE123'
    const request = getMockRequest(referenceNumber)
    const responseToolkit = getMockResponseToolkit()

    await permissionRenewalData[0].options.handler(request, responseToolkit)

    expect(permissionForFullReferenceNumber).toHaveBeenCalledWith(referenceNumber)
  })

  it('should call preparePermissionDataForRenewal with the expected data', async () => {
    const permissionMock = permissionForFullReferenceNumberMock()
    executeQuery.mockResolvedValueOnce([permissionMock])

    await permissionRenewalData[0].options.handler(getMockRequest(), getMockResponseToolkit())

    const expectedData = {
      ...permissionMock.entity.toJSON(),
      licensee: permissionMock.expanded.licensee.entity.toJSON(),
      concessions: [],
      permit: permissionMock.expanded.permit.entity.toJSON()
    }
    expect(preparePermissionDataForRenewal).toHaveBeenCalledWith(expectedData)
  })

  it('should return continue response', async () => {
    executeQuery.mockResolvedValueOnce([permissionForFullReferenceNumberMock()])
    const request = getMockRequest({})
    const responseToolkit = getMockResponseToolkit()

    expect(await permissionRenewalData[0].options.handler(request, responseToolkit)).toEqual(responseToolkit.continue)
  })

  describe('if there are no matching permissions', () => {
    it('should raise an error', async () => {
      executeQuery.mockResolvedValueOnce([])
      const request = getMockRequest({})
      const responseToolkit = getMockResponseToolkit()

      const boomError = Boom.unauthorized('Permission not found for renewal')
      await expect(permissionRenewalData[0].options.handler(request, responseToolkit)).rejects.toThrow(boomError)
    })
  })

  describe('if there are multiple matching permissions', () => {
    it('should raise an error', async () => {
      executeQuery.mockResolvedValueOnce(['foo', 'bar'])
      const request = getMockRequest({})
      const responseToolkit = getMockResponseToolkit()

      const error = 'Unable to get permission data for renewal, non-unique results for query'
      await expect(permissionRenewalData[0].options.handler(request, responseToolkit)).rejects.toThrow(error)
    })
  })

  it('should throw error if query execution fails', async () => {
    permissionForFullReferenceNumber.mockResolvedValueOnce({ filter: 'coffee' })
    const error = 'Wuh-oh!'
    executeQuery.mockImplementationOnce(() => {
      throw new Error(error)
    })
    const request = getMockRequest({})
    const responseToolkit = getMockResponseToolkit()

    await expect(permissionRenewalData[0].options.handler(request, responseToolkit)).rejects.toThrow(error)
  })
})
