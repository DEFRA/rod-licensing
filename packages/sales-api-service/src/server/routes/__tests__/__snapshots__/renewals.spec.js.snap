// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permissionRenewalData metadata should have expected description 1`] = `"Prepare data for renewing a permission based on the existing data"`;

exports[`permissionRenewalData metadata should have expected notes 1`] = `
"
        Prepare data for renewing a permission based on the existing data
      "
`;

exports[`permissionRenewalData metadata should have expected tags 1`] = `
Array [
  "api",
  "renewals",
]
`;

exports[`permissionRenewalData metadata should have expected validate 1`] = `
Object {
  "params": Object {
    "$_root": Object {
      "ValidationError": [Function],
      "_types": Set {
        "alternatives",
        "any",
        "array",
        "boolean",
        "date",
        "function",
        "link",
        "number",
        "object",
        "string",
        "symbol",
        "binary",
      },
      "allow": [Function],
      "alt": [Function],
      "alternatives": [Function],
      "any": [Function],
      "array": [Function],
      "assert": [Function],
      "attempt": [Function],
      "binary": [Function],
      "bool": [Function],
      "boolean": [Function],
      "build": [Function],
      "cache": Object {
        "provision": [Function],
      },
      "checkPreferences": [Function],
      "compile": [Function],
      "custom": [Function],
      "date": [Function],
      "defaults": [Function],
      "disallow": [Function],
      "equal": [Function],
      "exist": [Function],
      "expression": [Function],
      "extend": [Function],
      "forbidden": [Function],
      "func": [Function],
      "function": [Function],
      "in": [Function],
      "invalid": [Function],
      "isError": [Function],
      "isExpression": [Function],
      "isRef": [Function],
      "isSchema": [Function],
      "link": [Function],
      "not": [Function],
      "number": [Function],
      "object": [Function],
      "only": [Function],
      "optional": [Function],
      "options": [Function],
      "override": Symbol(override),
      "preferences": [Function],
      "prefs": [Function],
      "ref": [Function],
      "required": [Function],
      "string": [Function],
      "strip": [Function],
      "symbol": [Function],
      "trace": [Function],
      "types": [Function],
      "untrace": [Function],
      "valid": [Function],
      "version": "17.6.0",
      "when": [Function],
      "x": [Function],
      Symbol(@hapi/lab/coverage/initialize): [Function],
    },
    "$_super": Object {
      "default": [Function],
    },
    "$_temp": Object {
      "ruleset": false,
      "whens": Object {},
    },
    "$_terms": Object {
      "alterations": null,
      "dependencies": null,
      "examples": null,
      "externals": null,
      "keys":  [
        Object {
          "key": "referenceNumber",
          "schema": Object {
            "$_root": Object {
              "ValidationError": [Function],
              "_types": Set {
                "alternatives",
                "any",
                "array",
                "boolean",
                "date",
                "function",
                "link",
                "number",
                "object",
                "string",
                "symbol",
                "binary",
              },
              "allow": [Function],
              "alt": [Function],
              "alternatives": [Function],
              "any": [Function],
              "array": [Function],
              "assert": [Function],
              "attempt": [Function],
              "binary": [Function],
              "bool": [Function],
              "boolean": [Function],
              "build": [Function],
              "cache": Object {
                "provision": [Function],
              },
              "checkPreferences": [Function],
              "compile": [Function],
              "custom": [Function],
              "date": [Function],
              "defaults": [Function],
              "disallow": [Function],
              "equal": [Function],
              "exist": [Function],
              "expression": [Function],
              "extend": [Function],
              "forbidden": [Function],
              "func": [Function],
              "function": [Function],
              "in": [Function],
              "invalid": [Function],
              "isError": [Function],
              "isExpression": [Function],
              "isRef": [Function],
              "isSchema": [Function],
              "link": [Function],
              "not": [Function],
              "number": [Function],
              "object": [Function],
              "only": [Function],
              "optional": [Function],
              "options": [Function],
              "override": Symbol(override),
              "preferences": [Function],
              "prefs": [Function],
              "ref": [Function],
              "required": [Function],
              "string": [Function],
              "strip": [Function],
              "symbol": [Function],
              "trace": [Function],
              "types": [Function],
              "untrace": [Function],
              "valid": [Function],
              "version": "17.6.0",
              "when": [Function],
              "x": [Function],
              Symbol(@hapi/lab/coverage/initialize): [Function],
            },
            "$_super": Object {},
            "$_temp": Object {
              "ruleset": false,
              "whens": Object {},
            },
            "$_terms": Object {
              "alterations": null,
              "examples": Array [
                "17030621-3WC3FFT-B6HLG9",
              ],
              "externals": null,
              "metas": Array [],
              "notes": Array [],
              "replacements": null,
              "shared": null,
              "tags": Array [],
              "whens": null,
            },
            "_cache": null,
            "_flags": Object {
              "description": "The permission reference number",
              "presence": "required",
            },
            "_ids": Object {
              "_byId": Map {},
              "_byKey": Map {},
              "_schemaChain": false,
            },
            "_invalids": null,
            "_preferences": null,
            "_refs": Object {
              "refs": Array [],
            },
            "_rules": Array [
              Object {
                "_resolve": Array [],
                "args": Object {
                  "enabled": true,
                },
                "method": "trim",
                "name": "trim",
              },
              Object {
                "_resolve": Array [],
                "args": Object {
                  "direction": "upper",
                },
                "method": "case",
                "name": "case",
              },
              Object {
                "_resolve": Array [],
                "args": Object {
                  "options": Object {},
                  "regex": /\\^\\\\d\\{8\\}-\\\\d\\[A-Z\\]\\{2\\}\\\\d\\[A-Z\\]\\{3\\}-\\[A-Z0-9\\]\\{6\\}\\$/,
                },
                "errorCode": "string.pattern.base",
                "method": "pattern",
                "name": "pattern",
              },
            ],
            "_singleRules": Map {
              "trim" => Object {
                "_resolve": Array [],
                "args": Object {
                  "enabled": true,
                },
                "method": "trim",
                "name": "trim",
              },
              "case" => Object {
                "_resolve": Array [],
                "args": Object {
                  "direction": "upper",
                },
                "method": "case",
                "name": "case",
              },
            },
            "_valids": null,
            "type": "string",
          },
        },
      ],
      "metas": Array [],
      "notes": Array [],
      "patterns": null,
      "renames": null,
      "shared": null,
      "tags": Array [],
      "whens": null,
    },
    "_cache": null,
    "_flags": Object {
      "label": "permission-renewal-data-request-params",
    },
    "_ids": Object {
      "_byId": Map {},
      "_byKey": Map {
        "referenceNumber" => Object {
          "id": "referenceNumber",
          "schema": Object {
            "$_root": Object {
              "ValidationError": [Function],
              "_types": Set {
                "alternatives",
                "any",
                "array",
                "boolean",
                "date",
                "function",
                "link",
                "number",
                "object",
                "string",
                "symbol",
                "binary",
              },
              "allow": [Function],
              "alt": [Function],
              "alternatives": [Function],
              "any": [Function],
              "array": [Function],
              "assert": [Function],
              "attempt": [Function],
              "binary": [Function],
              "bool": [Function],
              "boolean": [Function],
              "build": [Function],
              "cache": Object {
                "provision": [Function],
              },
              "checkPreferences": [Function],
              "compile": [Function],
              "custom": [Function],
              "date": [Function],
              "defaults": [Function],
              "disallow": [Function],
              "equal": [Function],
              "exist": [Function],
              "expression": [Function],
              "extend": [Function],
              "forbidden": [Function],
              "func": [Function],
              "function": [Function],
              "in": [Function],
              "invalid": [Function],
              "isError": [Function],
              "isExpression": [Function],
              "isRef": [Function],
              "isSchema": [Function],
              "link": [Function],
              "not": [Function],
              "number": [Function],
              "object": [Function],
              "only": [Function],
              "optional": [Function],
              "options": [Function],
              "override": Symbol(override),
              "preferences": [Function],
              "prefs": [Function],
              "ref": [Function],
              "required": [Function],
              "string": [Function],
              "strip": [Function],
              "symbol": [Function],
              "trace": [Function],
              "types": [Function],
              "untrace": [Function],
              "valid": [Function],
              "version": "17.6.0",
              "when": [Function],
              "x": [Function],
              Symbol(@hapi/lab/coverage/initialize): [Function],
            },
            "$_super": Object {},
            "$_temp": Object {
              "ruleset": false,
              "whens": Object {},
            },
            "$_terms": Object {
              "alterations": null,
              "examples": Array [
                "17030621-3WC3FFT-B6HLG9",
              ],
              "externals": null,
              "metas": Array [],
              "notes": Array [],
              "replacements": null,
              "shared": null,
              "tags": Array [],
              "whens": null,
            },
            "_cache": null,
            "_flags": Object {
              "description": "The permission reference number",
              "presence": "required",
            },
            "_ids": Object {
              "_byId": Map {},
              "_byKey": Map {},
              "_schemaChain": false,
            },
            "_invalids": null,
            "_preferences": null,
            "_refs": Object {
              "refs": Array [],
            },
            "_rules": Array [
              Object {
                "_resolve": Array [],
                "args": Object {
                  "enabled": true,
                },
                "method": "trim",
                "name": "trim",
              },
              Object {
                "_resolve": Array [],
                "args": Object {
                  "direction": "upper",
                },
                "method": "case",
                "name": "case",
              },
              Object {
                "_resolve": Array [],
                "args": Object {
                  "options": Object {},
                  "regex": /\\^\\\\d\\{8\\}-\\\\d\\[A-Z\\]\\{2\\}\\\\d\\[A-Z\\]\\{3\\}-\\[A-Z0-9\\]\\{6\\}\\$/,
                },
                "errorCode": "string.pattern.base",
                "method": "pattern",
                "name": "pattern",
              },
            ],
            "_singleRules": Map {
              "trim" => Object {
                "_resolve": Array [],
                "args": Object {
                  "enabled": true,
                },
                "method": "trim",
                "name": "trim",
              },
              "case" => Object {
                "_resolve": Array [],
                "args": Object {
                  "direction": "upper",
                },
                "method": "case",
                "name": "case",
              },
            },
            "_valids": null,
            "type": "string",
          },
        },
      },
      "_schemaChain": false,
    },
    "_invalids": null,
    "_preferences": null,
    "_refs": Object {
      "refs": Array [],
    },
    "_rules": Array [],
    "_singleRules": Map {},
    "_valids": null,
    "type": "object",
  },
}
`;
